File: src/org/simbrain/workspace/updater/WorkspaceUpdater.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.workspace.updater;

import org.apache.log4j.Logger;
import org.simbrain.workspace.Workspace;
import org.simbrain.workspace.WorkspaceComponent;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * This class manages workspace updates. "Running" and "Stepping" the simulation
 * happen here, in a way that allows for concurrent update (in some cases) and
 * also interacts properly with single threaded guis using a "task
 * synchronization manager". Notification events about workspace events are
 * fired from here. Every time the workspace is updated, a list of actions in
 * the UpdateActionManager is invoked. By default one single action, a "buffered
 * update", occurs, in which components are updated in parallel, and when they
 * have all finished updating, couplings are updated. The update action manager
 * can also be used to customize update. Three executor services are here, one
 * for workspace updates (a single thread), one for event notification updates
 * (a single thread), and one for component updates (a thread pool with multiple
 * threads that can be configured), for cases when component updating happens
 * concurrently.
 *
 * @author Matt Watson
 * @author Jeff Yoshimi
 */
public class WorkspaceUpdater {

    /**
     * The static logger for the class.
     */
    static final Logger LOGGER = Logger.getLogger(WorkspaceUpdater.class);

    /**
     * The parent workspace.
     */
    private final Workspace workspace;

    /**
     * The executor service for managing workspace updates.
     */
    private final ExecutorService workspaceUpdateExecutor;

    /**
     * The executor service for notifying listeners.
     */
    private final ExecutorService notificationEvents;

    /**
     * Component listeners.
     */
    private final List<UpdateEventListener> componentListeners = new CopyOnWriteArrayList<UpdateEventListener>();

    /**
     * Updater listeners.
     */
    private final List<WorkspaceUpdaterListener> updaterListeners = new CopyOnWriteArrayList<WorkspaceUpdaterListener>();

    /**
     * Creates a default synch-manager that does nothing.
     */
    private volatile TaskSynchronizationManager syncManager = NO_ACTION_SYNC_MANAGER;

    /**
     * Executes queued InvocationEvents (i.e. AWT-driven events).
     */
    private SynchronizedTaskUpdateAction syncUpdateAction = new SynchronizedTaskUpdateAction(syncManager);

    /**
     * Whether updates should continue to run.
     */
    private volatile boolean run = false;

    /**
     * The number of times the update has run.
     */
    private volatile int time;

    /**
     * Number of threads used in the update service.
     */
    private int numThreads;

    /**
     * The update Manager.
     */
    private UpdateActionManager updateActionManager;

    /**
     * Constructor for the updater that uses the provided controller and
     * threads.
     *
     * @param workspace The parent workspace.
     * @param threads   The number of threads for component updates.
     */
    public WorkspaceUpdater(Workspace workspace, int threads) {
        this.workspace = workspace;
        this.numThreads = threads;
        // A single thread updates the workspace
        workspaceUpdateExecutor = Executors.newSingleThreadExecutor();
        // A single thread to fire notification events
        notificationEvents = Executors.newSingleThreadExecutor();
        // Instantiate the update action manager
        updateActionManager = new UpdateActionManager(this);
    }

    /**
     * Constructor for the updater that uses the default controller and default
     * number of threads.
     *
     * @param workspace The parent workspace.
     */
    public WorkspaceUpdater(Workspace workspace) {
        this(workspace, Runtime.getRuntime().availableProcessors());
    }

    /**
     * Sets the manager. Setting the manager to null clears the manager.
     *
     * @param manager the new manager.
     */
    public void setTaskSynchronizationManager(TaskSynchronizationManager manager) {
        if (manager == null) {
            syncManager = NO_ACTION_SYNC_MANAGER;
        } else {
            syncManager = manager;
        }
        syncUpdateAction.setSyncManager(syncManager);
    }

    /**
     * Returns the update action used to execute synchronized invocation events. This should be added to the update
     * action manager if the update sequence is cleared or deserialized.
     */
    public UpdateAction getSyncUpdateAction() {
        return syncUpdateAction;
    }

    /**
     * Returns the 'time' or number of update iterations that have passed.
     *
     * @return The time.
     */
    public int getTime() {
        return time;
    }

    /**
     * Sets the time.
     *
     * @param time time to set
     */
    public void setTime(final int time) {
        this.time = time;
    }

    /**
     * Reset time to 0.
     */
    public void resetTime() {
        time = 0;
    }

    /**
     * Stops the update thread.
     */
    public void stop() {
        run = false;
    }

    /**
     * Returns whether the updater is set to run.
     */
    public boolean isRunning() {
        return run;
    }

    /**
     * Starts the update thread. Used when "running" the workspace by pressing
     * the play button in the gui.
     */
    public void run() {
        run = true;
        for (WorkspaceComponent wc : workspace.getComponentList()) {
            wc.setRunning(true);
        }
        workspaceUpdateExecutor.submit(() -> {
            notifyWorkspaceUpdateStarted();
            syncManager.queueTasks();
            try {
                while (run) {
                    doUpdate();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            run = false;
            for (WorkspaceComponent component : workspace.getComponentList()) {
                component.setRunning(false);
            }
            syncManager.releaseTasks();
            syncManager.runTasks();
            notifyWorkspaceUpdateCompleted();
        });
    }

    /**
     * Submits a single task to the queue.
     */
    public void runOnce() {
        workspaceUpdateExecutor.submit(() -> {
            notifyWorkspaceUpdateStarted();
            syncManager.queueTasks();
            try {
                doUpdate();
            } catch (Exception e) {
                e.printStackTrace();
            }
            syncManager.releaseTasks();
            syncManager.runTasks();
            notifyWorkspaceUpdateCompleted();
        });
    }

    /**
     * Iterate a set number of iterations.
     * <p>
     * See {@link Workspace#iterate(int)}
     *
     * @param numIterations the number of iterations to update
     */
    public void iterate(int numIterations) {
        Future<?> wait = workspaceUpdateExecutor.submit(() -> {
            notifyWorkspaceUpdateStarted();
            try {
                syncManager.queueTasks();
                for (int i = 0; i < numIterations; i++) {
                    doUpdate();
                }
                syncManager.releaseTasks();
                syncManager.runTasks();
            } catch (Exception e) {
                e.printStackTrace();
            }
            notifyWorkspaceUpdateCompleted();
        });
        try {
            wait.get();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Executes the main workspace update.
     */
    private void doUpdate() {
        time++;
        LOGGER.trace("starting: " + time);
        for (UpdateAction action : updateActionManager.getActionList()) {
            notifyBeforeUpdateAction(action);
            action.invoke();
            notifyAfterUpdateAction(action);
        }
        notifyWorkspaceUpdated();
        LOGGER.trace("done: " + time);
    }

    /**
     * Adds a component listener to this instance.
     *
     * @param listener The component listener to add.
     */
    public void addComponentListener(UpdateEventListener listener) {
        componentListeners.add(listener);
    }

    /**
     * Removes a component listener from this instance.
     *
     * @param listener The listener to add.
     */
    public void removeComponentListener(UpdateEventListener listener) {
        componentListeners.remove(listener);
    }

    /**
     * Adds an updater listener to this instance.
     *
     * @param listener updater component listener to add.
     */
    public void addUpdaterListener(WorkspaceUpdaterListener listener) {
        updaterListeners.add(listener);
    }

    /**
     * Removes an updater listener from this instance.
     *
     * @param listener The updater listener to add.
     */
    public void removeUpdaterListener(WorkspaceUpdaterListener listener) {
        updaterListeners.remove(listener);
    }

    /**
     * Called when an update action is about to be invoked.
     *
     * @param action The action to be invoked.
     */
    void notifyBeforeUpdateAction(UpdateAction action) {
        final long nanoTime = System.nanoTime();
        notificationEvents.submit(() -> {
            componentListeners.forEach(l -> l.beforeUpdateAction(action, nanoTime));
        });
    }

    /**
     * Called after an update action has been invoked.
     *
     * @param action The action that was invoked.
     */
    void notifyAfterUpdateAction(UpdateAction action) {
        final long nanoTime = System.nanoTime();
        notificationEvents.submit(() -> {
            componentListeners.forEach(l -> l.afterUpdateAction(action, nanoTime));
        });
    }

    /**
     * Called when a new component is starting to update.
     *
     * @param component The component to update.
     * @param thread    The number of the thread doing the update.
     */
    void notifyComponentUpdateStarted(WorkspaceComponent component, int thread) {
        final int simTime = this.time;
        final long nanoTime = System.nanoTime();
        notificationEvents.submit(() -> {
            componentListeners.forEach(l -> l.beforeComponentUpdate(component, simTime, thread, nanoTime));
        });
    }

    /**
     * Called when a new component is finished updating.
     *
     * @param component The component to update.
     * @param thread    The number of the thread doing the update.
     */
    void notifyComponentUpdateFinished(WorkspaceComponent component, int thread) {
        final int simTime = this.time;
        final long nanoTime = System.nanoTime();
        notificationEvents.submit(() -> {
            componentListeners.forEach(l -> l.afterComponentUpdate(component, simTime, thread, nanoTime));
        });
    }

    /**
     * Called when the couplings are updated.
     */
    protected void notifyCouplingsUpdated() {
        final int time = this.time;
        Consumer<WorkspaceUpdaterListener> notifier = l -> l.updatedCouplings(time);
        notificationEvents.submit(() -> updaterListeners.forEach(notifier));
    }

    /**
     * Called when the workspace update begins.
     */
    private void notifyWorkspaceUpdateStarted() {
        notificationEvents.submit(() -> updaterListeners.forEach(WorkspaceUpdaterListener::updatingStarted));
    }

    /**
     * Called when workspace update finishes.
     */
    private void notifyWorkspaceUpdateCompleted() {
        notificationEvents.submit(() -> updaterListeners.forEach(WorkspaceUpdaterListener::updatingFinished));
    }

    /**
     * Called after every workspace update .
     */
    private void notifyWorkspaceUpdated() {
        notificationEvents.submit(() -> updaterListeners.forEach(WorkspaceUpdaterListener::workspaceUpdated));
    }

    /**
     * @return the numThreads
     */
    public int getNumThreads() {
        return numThreads;
    }

    /**
     * Set number of threads in updater.
     *
     * @param numThreads number of threads.
     */
    public void setNumThreads(final int numThreads) {
        if (isRunning()) {
            stop();
        }
        this.numThreads = numThreads;
        // this.componentUpdates = Executors.newFixedThreadPool(numThreads,
        // new UpdaterThreadFactory());
        for (WorkspaceUpdaterListener listener : updaterListeners) {
            listener.changeNumThreads();
        }

    }

    /**
     * A synch-manager where the methods do nothing.
     */
    private static final TaskSynchronizationManager NO_ACTION_SYNC_MANAGER = new TaskSynchronizationManager() {
        public void queueTasks() {
            /* no implementation */
        }

        public void releaseTasks() {
            /* no implementation */
        }

        public void runTasks() {
            /* no implementation */
        }
    };

    /**
     * Returns a reference to the update manager.
     *
     * @return the update manager
     */
    public UpdateActionManager getUpdateManager() {
        return updateActionManager;
    }

    /**
     * Get a synchronized list of component.
     *
     * @return the synchronized list of components
     */
    public List<? extends WorkspaceComponent> getComponents() {
        List<? extends WorkspaceComponent> components = workspace.getComponentList();
        synchronized (components) {
            components = new ArrayList<WorkspaceComponent>(components);
        }
        return components;
    }

    /**
     * @return the workspace
     */
    public Workspace getWorkspace() {
        return workspace;
    }

}

File: src/main/kotlin/org/simbrain/workspace/updater/WorkspaceUpdater.kt
package org.simbrain.workspace.updater

import kotlinx.coroutines.*
import org.pmw.tinylog.Logger
import java.util.concurrent.ConcurrentHashMap
import javax.swing.SwingUtilities
import org.simbrain.workspace.Workspace
import org.simbrain.workspace.WorkspaceComponent
import org.simbrain.workspace.events.WorkspaceUpdaterEvents

/**
 * This class manages workspace updates. "Running" and "Stepping" the simulation
 * happen here, in a way that allows for concurrent update (in some cases) and
 * also interacts properly with single threaded guis using a "task
 * synchronization manager". Notification events about workspace events are
 * fired from here. Every time the workspace is updated, a list of actions in
 * the UpdateActionManager is invoked. By default one single action, a "buffered
 * update", occurs, in which components are updated in parallel, and when they
 * have all finished updating, couplings are updated. The update action manager
 * can also be used to customize update. Three executor services are here, one
 * for workspace updates (a single thread), one for event notification updates
 * (a single thread), and one for component updates (a thread pool with multiple
 * threads that can be configured), for cases when component updating happens
 * concurrently.
 *
 * To make a new custom UpdateAction in kotlin use the updateAction() function.
 *
 * @author Matt Watson
 * @author Jeff Yoshimi
 */
class WorkspaceUpdater(val workspace: Workspace) {

    val events = WorkspaceUpdaterEvents()

    /**
     * Whether updates should continue to run.
     */
    var isRunning = false
        private set

    /**
     * The number of times the update has run.
     */
    var time = 0

    /**
     * The update Manager.
     */
    val updateManager: UpdateActionManager = UpdateActionManager(this)

    /**
     * The jobs of every iteration currently inside [doUpdate] — entry points can overlap, e.g. a
     * running loop plus a GUI-triggered single iterate — so a stop can escalate to cancelling them
     * when a suspend attribute or update action hangs instead of coming back to the loop's
     * [isRunning] check.
     */
    private val activeIterationJobs: MutableSet<Job> = ConcurrentHashMap.newKeySet()

    /**
     * Whether some iteration is inside [doUpdate] right now. With [isRunning] already false, this is
     * the signal a stop request did not come back — the GUI stop action escalates to [stopNow] on a
     * repeated press in that state.
     */
    val hasActiveIteration: Boolean
        get() = activeIterationJobs.isNotEmpty()

    /**
     * Reset time to 0.
     */
    fun resetTime() {
        time = 0
    }

    /**
     * Requests a cooperative stop: the current iteration finishes and the loop exits, which keeps the
     * step deterministic. Safe to call redundantly — programmatic callers do — so escalation is not
     * automatic here; a caller that knows the user wants out, such as the stop button on a repeated
     * press, uses [stopNow].
     */
    fun stop() {
        isRunning = false
    }

    /**
     * Stops and also cancels the in-flight iterations cooperatively, abandoning whatever step was in
     * progress. Suspend attributes are cancelled at their suspension points; blocking work must watch
     * its own abort flag.
     */
    fun stopNow() {
        isRunning = false
        activeIterationJobs.toList().forEach { it.cancel() }
    }

    /**
     * Starts the update thread. Used when "running" the workspace by pressing
     * the play button in the gui.
     */
    suspend fun run() = coroutineScope {
        isRunning = true
        for (wc in workspace.componentList) {
            wc.isRunning = true
        }
        events.runStarted.fire()
        try {
            withContext(workspace.coroutineContext) {
                while (isRunning) {
                    doUpdate()
                }
            }
        } catch (e: CancellationException) {
            // Rethrows when this caller itself was cancelled; a cancellation from [stopNow] hitting the
            // in-flight iteration instead ends the run normally
            currentCoroutineContext().ensureActive()
        } finally {
            isRunning = false
            for (component in workspace.componentList) {
                component.isRunning = false
            }
            events.runFinished.fire()
        }
    }

    /**
     * Submits a single task to the queue.
     */
    suspend fun runOnce() {
        isRunning = true
        for (wc in workspace.componentList) {
            wc.isRunning = true
        }
        events.runStarted.fire()
        try {
            withContext(workspace.coroutineContext) {
                doUpdate()
            }
        } catch (e: CancellationException) {
            currentCoroutineContext().ensureActive()
        } finally {
            events.runFinished.fire()
            isRunning = false
            for (component in workspace.componentList) {
                component.isRunning = false
            }
        }
    }

    fun runBlocking() {
        assertNotOnEventThread()
        isRunning = true
        for (wc in workspace.componentList) {
            wc.isRunning = true
        }
        runBlocking {
            events.runStarted.fire()
            try {
                doUpdate()
            } catch (e: CancellationException) {
                coroutineContext.ensureActive()
            }
            events.runFinished.fire()
        }
        isRunning = false
        for (component in workspace.componentList) {
            component.isRunning = false
        }
    }

    /**
     * Iterate a set number of iterations.
     *
     * Optional finishing task is run after main iteration finishes.
     *
     * See [Workspace.iterateAsync]
     *
     * @param numIterations the number of iterations to update
     */
    suspend fun iterate(numIterations: Int, finishingTask: () -> Unit = {}) {
        isRunning = true
        for (wc in workspace.componentList) {
            wc.isRunning = true
        }
        events.runStarted.fire()
        try {
            repeat(numIterations) {
                doUpdate()
            }
        } catch (e: CancellationException) {
            currentCoroutineContext().ensureActive()
        } finally {
            isRunning = false
            finishingTask()
            for (component in workspace.componentList) {
                component.isRunning = false
            }
            events.runFinished.fire()
        }
    }

    suspend fun iterateWhile(predicate: () -> Boolean) {
        isRunning = true
        for (wc in workspace.componentList) {
            wc.isRunning = true
        }
        events.runStarted.fireAsync()
        try {
            do {
                doUpdate()
            } while (predicate())
        } catch (e: CancellationException) {
            currentCoroutineContext().ensureActive()
        } finally {
            isRunning = false
            for (component in workspace.componentList) {
                component.isRunning = false
            }
            events.runFinished.fireAsync()
        }
    }

    /**
     * Executes the main workspace update. The iteration's job is exposed to [stopNow] for the duration,
     * so a stuck iteration can be cancelled from outside.
     */
    private suspend fun doUpdate() {
        time++
        Logger.trace("starting: $time")
        withContext(workspace.coroutineContext) {
            val iterationJob = coroutineContext.job
            activeIterationJobs.add(iterationJob)
            try {
                for (action in updateManager.actionList + updateManager.nonRemovableActions) {
                    with(PerformanceMonitor) {
                        action()
                    }
                }
            } finally {
                activeIterationJobs.remove(iterationJob)
            }
        }
        events.workspaceUpdated.fire()
        Logger.trace("done: $time")
    }

    private fun assertNotOnEventThread() {
        check(!SwingUtilities.isEventDispatchThread()) {
            "Blocking workspace iteration must not run on the event dispatch thread: suspend " +
                    "consumables wait for event thread handlers, which deadlocks against a blocked " +
                    "event thread. Use a launched iteration (e.g. Workspace.iterateAsync) instead."
        }
    }

    init {
        // A single thread updates the workspace
        // A single thread to fire notification events
        // Instantiate the update action manager
    }

    /**
     * Get a synchronized list of component.
     *
     * @return the synchronized list of components
     */
    val components: List<WorkspaceComponent>
        get() {
            var components = workspace.componentList
            synchronized(components) { components = ArrayList(components) }
            return components
        }

}
